services/src/datasets/external/aruna/mod.rs

```diff
@@ -212,106 +217,299 @@ impl ArunaDataProvider {
 
-        if aruna_object_group_overview.len() != 1 {
-            return Err(ArunaProviderError::UnexpectedObjectHierarchy);
+        for dataset in requested_datasets {
+            if self
+                .verify_resource(&dataset.id, &dataset.key_values, dataset.status())
+                .is_ok()
+            {
+                let outgoing_objects = Self::get_outgoing_internal_relation_ids(
+                    &dataset.relations,
+                    ResourceVariant::Object,
+                )
+                .ok();
+                if let Some(object_ids) = outgoing_objects {
+                    if self.get_verified_dataset_objects(object_ids).await.is_ok() {
+                        verified_datasets.push(dataset);
+                    }
+                }
+            }
         }
-        let object_group_id = aruna_object_group_overview
-            .pop()
-            .expect("Object groups should have size one")
-            .id;
-
-        let aruna_objects = object_group_stub
-            .get_object_group_objects(GetObjectGroupObjectsRequest {
-                collection_id: collection_id.clone(),
-                group_id: object_group_id.clone(),
-                page_request: None,
-                meta_only: false,
-            })
+        // //TODO: Workaround until deleted access permissions are fixed.
+        // let mut datasets = vec![];
+        //
+        // for dataset_id in dataset_ids {
+        //     let dataset_response = dataset_stub
+        //         .get_dataset(GetDatasetRequest {
+        //             dataset_id: dataset_id.clone(),
+        //         })
+        //         .await?;
+        //
+        //     if dataset_response.is_err_and(|x| x.code() == Code::Unauthenticated) {
+        //         log::debug!(
+        //             "Ignoring Code::Unauthenticated Error in Aruna Dataset Request For id={} (Check if Dataset is deleted)",
+        //             dataset_id,
+        //         )
+        //     } else {
+        //         let dataset = dataset_response
+        //             .into_inner()
+        //             .dataset
+        //             .ok_or(ArunaProviderError::MissingDataset)?;
+        //         if dataset.status() == aruna_rust_api::api::storage::models::v2::Status::Available
+        //             && self.has_filter_label(&dataset.key_values)
+        //         {
+        //             datasets.push(dataset);
+        //         }
+        //     }
+        // }
+
+        Ok(verified_datasets)
+    }
+
+    async fn get_verified_dataset_objects(
+        &self,
+        object_ids: Vec<String>,
+    ) -> Result<ArunaDatasetObjectIds> {
+        let mut object_stub = self.object_stub.clone();
+
+        let requested_objects = object_stub
+            .get_objects(GetObjectsRequest { object_ids })
             .await?
             .into_inner()
-            .object_group_objects;
+            .objects;
+
+        let mut verified_objects = vec![];
```

Comment:
why do you collect the objects in the client? Is it not possible to filter in the server?

Reply:
The Object is just the metadata of the object (i.e., name, description, labels, ...), not the object itself. 

There is a [Search API](https://arunastorage.github.io/documentation/latest/get_started/basic_usage/12_How-To-Search) in Aruna, but I saw two problems with it based on our current Aruna implementation. Our current Aruna implementation assumes there is a Geo Engine project in Aruna that contains the datasets that should be visualized in Geo Engine for which we have an API key. The Search API queries all public objects.

(1) The Geo Engine project and its object might be private, or classified, or whatever. Just having public objects seems limiting. Public objects might also contain some sort of weird relationships between objects that we cannot properly handle.
(2) The documentation of the Search API does not specify fields to query the relation between object, i.e. this meta object belongs to that data object or these objects belong to the Geo Engine project...so we would have to verify the results afterwards anyway.

If the querying of metadata about objects becomes a bottleneck and/or the Search API is changed and/or we change the way we handle Aruna access, I think this should be revisited in another pull request. For now, I chose the solution that mirrors our previous handling of Aruna to update to the new dependency.